File: python/packages/spore_eval/src/spore_eval/eval/manifest.py
"""Manifest loading + manual promotion (Rules 6, 29, 31).

Mirrors ``rust/crates/spore-eval/src/manifest.rs``.
"""

from __future__ import annotations

import json
from pathlib import Path

from pydantic import ValidationError

from .task import (
    ManifestParseError,
    MissingSuiteVersionError,
    TaskSuite,
)


def load_suite_str(body: str) -> TaskSuite:
    """Load a :class:`TaskSuite` from a JSON manifest string. Rejects a manifest
    without ``suite_version`` (Rule 6) with :class:`MissingSuiteVersionError`."""
    try:
        value = json.loads(body)
    except json.JSONDecodeError as e:
        raise ManifestParseError(f"manifest parse error: {e}") from e
    if not isinstance(value, dict) or "suite_version" not in value:
        raise MissingSuiteVersionError()
    try:
        return TaskSuite.model_validate(value)
    except ValidationError as e:
        raise ManifestParseError(f"manifest parse error: {e}") from e


def load_suite_path(path: Path | str) -> TaskSuite:
    """Load a :class:`TaskSuite` from a manifest file path."""
    return load_suite_str(Path(path).read_text(encoding="utf-8"))


def suite_to_json(suite: TaskSuite) -> str:
    """Serialize a :class:`TaskSuite` back to pretty JSON, omitting unset
    optional fields so the round-trip stays close to the source manifest."""
    return json.dumps(suite.model_dump(mode="json", exclude_none=True), indent=2)


def promote_challenge_task(suite: TaskSuite, task_id: str) -> None:
    """Manually promote a ``challenge`` task to ``regression``, bumping
    ``suite_version`` (Rule 31). Auto-promotion is deferred. Raises
    :class:`ManifestParseError` if ``task_id`` is not a challenge task."""
    pos = next((i for i, t in enumerate(suite.challenge) if t.id == task_id), None)
    if pos is None:
        raise ManifestParseError(f"challenge task {task_id!r} not found")
    task = suite.challenge.pop(pos)
    suite.regression.append(task)
    suite.suite_version += 1

File: go/spore-eval/manifest/manifest.go
// Package manifest handles task-suite manifest loading + manual promotion
// (Rules 6, 29, 31).
package manifest

import (
	"encoding/json"
	"fmt"
	"os"

	"github.com/squirrelsoft-dev/spore-core/go/spore-eval/task"
)

// LoadSuiteStr loads a TaskSuite from a JSON manifest string. It rejects a
// manifest without suite_version (Rule 6) with task.ErrMissingSuiteVersion.
func LoadSuiteStr(jsonStr string) (*task.TaskSuite, error) {
	// First check the raw JSON for the required field so a missing
	// suite_version is a precise error rather than a generic parse failure.
	var raw map[string]json.RawMessage
	if err := json.Unmarshal([]byte(jsonStr), &raw); err != nil {
		return nil, &task.ManifestParseError{Msg: err.Error()}
	}
	if _, ok := raw["suite_version"]; !ok {
		return nil, task.ErrMissingSuiteVersion
	}
	var suite task.TaskSuite
	if err := json.Unmarshal([]byte(jsonStr), &suite); err != nil {
		return nil, &task.ManifestParseError{Msg: err.Error()}
	}
	return &suite, nil
}

// LoadSuitePath loads a TaskSuite from a manifest file path.
func LoadSuitePath(path string) (*task.TaskSuite, error) {
	body, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	return LoadSuiteStr(string(body))
}

// SuiteToJSON serialises a TaskSuite back to pretty JSON.
func SuiteToJSON(suite *task.TaskSuite) (string, error) {
	b, err := json.MarshalIndent(suite, "", "  ")
	if err != nil {
		return "", &task.ManifestParseError{Msg: err.Error()}
	}
	return string(b), nil
}

// PromoteChallengeTask manually promotes a challenge task to regression,
// bumping suite_version (Rule 31). Auto-promotion is deferred. Returns an error
// if taskID is not found among the challenge tasks.
func PromoteChallengeTask(suite *task.TaskSuite, taskID string) error {
	pos := -1
	for i := range suite.Challenge {
		if string(suite.Challenge[i].ID) == taskID {
			pos = i
			break
		}
	}
	if pos < 0 {
		return &task.ManifestParseError{Msg: fmt.Sprintf("challenge task %q not found", taskID)}
	}
	t := suite.Challenge[pos]
	suite.Challenge = append(suite.Challenge[:pos], suite.Challenge[pos+1:]...)
	suite.Regression = append(suite.Regression, t)
	suite.SuiteVersion++
	return nil
}
